### PR TITLE
Refactor: Client Edge Case

### PR DESCRIPTION
**Pull request by whoisrey**
## 제목

클라이언트 코드 리팩토링

## 내용

Drag & Drop Edge Case를 해결하고 Test Code Dashboard를 개선하였습니다.

| Drag & Drop | Test Code Dashboard |
|-|-|
|  |  |

## 항목
- 블록 ID 값 부여 (`Date.now()` vs **UUID**)
  우리는 밀리초 단위로 난수를 생성하기 위해 간혹 `Date.now()`를 사용하지만 이 숫자는 항상 고유성을 보장하지는 않습니다. 
  > 컴퓨터는 밀리초 이내에도 몇 번의 루프를 반복하여 실행할 수 있습니다. 
  > 만약 여러 요소를 순회하는 반복문 안에서 각 요소에 id 값을 `Date.now()`로 부여할 경우, 
  > 같은 밀리초에 id값이 부여된 요소들은 같은 난수를 사용하는 <ins>위험한 상황</ins>이 발생할 수 있습니다.
  
  [참고자료](https://dev.to/rahmanfadhil/how-to-generate-unique-id-in-javascript-1b13)

- Drag & Drop Edge Case
  - 같은 라인 블록에서 개별 블록의 순서 이동
  - 다른 라인 블록으로 개별 블록의 순서 이동
  - 원하는 위치에 개별 블록 이동
 
- Test Code Dashboard
  - Test Code 생성할 때까지 Loading 화면 구현 (조건부 렌더링)
 
- 키보드로 블록을 제거하는 경우, 
  - Input Field를 수정할 때, 블록이 (의도와 다르게) 지워지는 상황 해결

## 특이 사항
- 추후 구현해야할 사항 8/23(금)

  - 블록이 이동하는 예상 위치를 표시 #24 
  - Test Code 전체 코드 복사 기능 #25

## 주의 사항

- [x] PR 크기는 300~500줄로 하되 최대 1000줄 미만으로 합니다.
- [x] conflict를 모두 해결하고 PR을 올려주세요

# PR 전 체크리스트

- [x] 가장 최신 브랜치를 pull 하였습니다
- [x] base 브랜치명을 확인하였습니다
- [x] 코드 컨벤션을 모두 확인하였습니다
- [x] 셀프 리뷰를 진행하였습니다.
- [x] 브랜치명을 확인하였습니다.
- [x] 작업 중 dependency 변경사항이 있는 경우 안내해주세요!

### CODE REVIEW

File: src/components/common/CodeBox.jsx [hunk 1]
```diff
@@ -1,10 +1,14 @@
 import { CodeBoxContainer, CodeBoxContent } from "../../style/CodeBoxStyle";
 
 const CodeBox = ({ testCode }) => {
+  const startText = "const { chromium } = require('playwright'); {";
```
Comment (choiOV) [inline src/components/common/CodeBox.jsx:4]: Test code Dashboard에 표시될 필수 테스트 코드를 하드코딩하여 보여주시도록 한 것으로 이해했습니다.
그런데 `startText`는 require문으로 보이는데 맨 마지막 `{`이 필요한 이유가 있을까요?

제가 잘못 이해한 것이라면 양해 부탁드립니다!🙏

File: src/store/index.jsx [hunk 2]
```diff
@@ -1,17 +1,18 @@
 import { create } from "zustand";
+import { v4 as uuidv4 } from "uuid";
 
 const useLineBlocksStore = create((set) => ({
-  lineBlocks: [{ id: Date.now(), blocks: [] }],
+  lineBlocks: [{ id: uuidv4(), blocks: [] }],
```
Comment (choiOV) [inline src/store/index.jsx:5]: 기존에는 고유한 id를 생성하기 위해 `Date` 객체를 사용하셨는데
혹시 모를 중복성을 방지하고자 안정된 `uuid`를 사용한 것이 맞을까요? 😊

File: src/store/index.jsx [hunk 3]
```diff
@@ -112,7 +161,14 @@ const useSelectionStore = create((set, get) => ({
       const { selectedBlockId } = get();
       const { lineBlocks, setLineBlocks } = useLineBlocksStore.getState();
 
-      if (selectedBlockId !== null) {
+      const isSelectedBlockId = selectedBlockId !== null;
+      const isTyping = ["INPUT"].includes(event.target.tagName);
+
+      if (isTyping) {
+        return;
+      }
```
Comment (choiOV) [inline src/store/index.jsx:169]: HTML 요소의 태그 이름을 검사해 사용자가 input 필드를 타이핑 중인지 확인하고,
타이핑 중이라면 블록 삭제 기능이 작동하지 않도록 만들어주셨군요! 👍

**Review by choiOV — APPROVED** [review]
블록의 ID 값을 UUID로 부여해 안정성을 증가시켰고,
Drag & Drop Edge Case, 블록 삭제 Edge Case 등을 멋지게 해결해주셨네요!

이번 코드 작업도 수고 많으셨습니다 양래님! 😊

File: src/components/common/CodeBox.jsx (hunk 1 above)
Comment (whoisrey) [reply src/components/common/CodeBox.jsx:4]: 맞습니다. 용섭님! 하드코딩으로 테스트 코드를 보여주려다보니 변수명의 사용의도와 맞지 않게 작성된 것 같습니다.
블럭 안에 테스트 코드를 넣기 위해 startText의 끝이 `{`로 시작했지만, 추후에 개선 작업 진행하겠습니다. 
#26

File: src/store/index.jsx (hunk 2 above)
Comment (whoisrey) [reply src/store/index.jsx:5]: 맞습니다. 용섭님! 밀리초 단위로 다른 난수를 생성하는 `Date.now()`는 같은 밀리초에 여러 작업을 실행할 수 있는 컴퓨터의 특성 상 고유한 번호를 생성하지 못할 위험이 있다고 판단하였습니다. `Date.now()`의 원래 목적은 고유 번호를 생성하는 것이 아니기도 하구요!

File: src/store/index.jsx (hunk 3 above)
Comment (whoisrey) [reply src/store/index.jsx:169]: 맞습니다. 용섭님! keyDown으로 삭제 기능을 구현하다보니 사용자가 input field를 수정할 때, 의도치 않게 블럭이 삭제되는 경우가 발생하였습니다. 이를 대처하고자 위와 같이 구현하였습니다!

File: src/store/index.jsx [hunk 4]
```diff
@@ -24,35 +25,78 @@ const useDragStore = create((set, get) => ({
...
 
+    const isSameLine = draggedLineIndex === targetLineIndex;
+
     if (draggedBlock !== null) {
       const newLineBlocks = lineBlocks.map((lineBlock, index) => {
-        const isDraggedLine = index === draggedLineIndex;
         const isTargetLine = index === targetLineIndex;
-        let newBlocks = [...lineBlock.blocks];
+        const isDraggedLine = index === draggedLineIndex;
+        const hasDraggedBlockId = draggedBlockId !== null;
```
Comment (noeyigg) [inline src/store/index.jsx:42]: uuid를 사용하시면서 id가 있는지 없는지 확인한 후 id를 생성하는 로직도 추가해주셨네요! 해당 조건을 hasDraggedBlockId라는 명확한 변수에 담아주셔서 이해하기에 수월했습니다!

File: src/store/index.jsx (hunk 3 above)
Comment (noeyigg) [reply src/store/index.jsx:169]: 덕분에 앞으로 keyDown으로 삭제 기능을 구현할 경우 타이핑중인지 확인하는 로직이 포함되어있는지 확인하는 습관을 들일 수 있을 것 같습니다!

**Review by noeyigg — COMMENTED** [review]
꼭 필요했던 개선사항이라는 생각이 듭니다! 꼼꼼하게 Edge Case를 확인하고 처리해주셨네요! 수고하셨습니다!

File: src/store/index.jsx (hunk 3 above)
Comment (whoisrey) [reply src/store/index.jsx:169]: 네 기연님! event.target.tagName에서 tag 이름은 항상 대문자가 온다는 것도 기억해주시면 좋을 것 같아요~
[참고자료](https://developer.mozilla.org/ko/docs/Web/API/Element/tagName)

File: src/store/index.jsx (hunk 4 above)
Comment (whoisrey) [reply src/store/index.jsx:42]: 감사합니다. 조건을 변수명에 담아서 로직을 작성하면 가독성이 좋아지는 것 같아요 기연님~
